Fix crossgenning of framework libraries on Windows

trylek opened the pull request:
I tried to slightly improve the data flow by creating explicit
environment variables for the two arguments. There were a bunch
of errors regarding % vs !, duplicate double-quoting and the like.
It seems to be working now.

Thanks

Tomas

trylek committed:
Fix crossgenning of framework libraries on Windows

I tried to slightly improve the data flow by creating explicit
environment variables for the two arguments. There were a bunch
of errors regarding % vs !, duplicate double-quoting and the like.
It seems to be working now.

Thanks

Tomas

janvorli commented on src/coreclr/build-test.cmd:
```diff
@@ -663,51 +663,65 @@ exit /b 0
+if /I "%AssemblyName%" == "mscorlib.ni.dll" exit /b 0
 REM don't precompile anything from CoreCLR
-if /I exist %CORE_ROOT_STAGE%\%2 exit /b 0
+if /I exist %CORE_ROOT_STAGE%\%AssemblyName% exit /b 0
```
Does this mean that we won't be crossgening System.Private.CoreLib.dll with crossgen2?

trylek replied:
It probably does and I concur it's something worth discussing moving forward just like the default compilation of S.P.C with legacy Crossgen within build.cmd/sh.

janvorli replied:
I think this is a bit different from using the new crossgen in build.cmd/sh. When we run the build-tests.cmd with the option asking for crossgen2 compilation, is it strange to get the whole framework crossgened with crossgen2 except for the S.P.C. I wonder if we could just add a check here that if we are using crossgen2, we would precompile the S.P.C. too.

trylek replied:
I agree it's weird. I basically left that in as a pre-existing conditional statement but thinking about this some more, what about throwing this away altogether? It doesn't make too much sense to me in the "new world".

janvorli replied:
That sounds reasonable.

trylek committed:
Enable crossgenning S.P.C with CG2, remove obsolete CoreCLR logic

janvorli reviewed (approved):
LGTM, thank you!

janvorli commented on src/coreclr/build-test.cmd:
```diff
@@ -663,51 +663,59 @@ exit /b 0
     set /a __exitCode = !errorlevel!
 )
 
+echo %__CrossgenCmd%
```
It seems this should be replaced  by the similar code from the `else` branch above and deleted from the `else` branch.

trylek replied:
Fixed, thanks for pointing that out!

janvorli commented on src/coreclr/build-test.cmd:
```diff
@@ -663,51 +663,59 @@ exit /b 0
+    set __CrossgenCmd=!__CrossgenExe! /Platform_Assemblies_Paths "!CORE_ROOT!" /in !AssemblyPath! /out !__CrossgenOutputFile! >nul 2>nul
 ) else (
-    "!CORE_ROOT_STAGE!\crossgen2\crossgen2 -r:"!CORE_ROOT!\*.dll" -O --inputbubble -out:"__CrossgenOutputFile" "%1" >nul 2>nul
+    set __CrossgenCmd=!__CrossgenExe! -r:"!CORE_ROOT!\*.dll" -O --inputbubble --out:!__CrossgenOutputFile! !AssemblyPath! >nul 2>nul
```
I think that to eliminate all the ugly warnings on bad image format of files that are .dll but not framework assemblies, we could replace `-r:"!CORE_ROOT!\*.dll` by `-r:"!CORE_ROOT!\System.*.dll -r:"!CORE_ROOT!\Windows.*.dll` I think this covers all the framework assemblies.

janvorli replied:
Oh, and `-r:!CORE_ROOT!\mscorlib.dll` too, some tests need that.

trylek replied:
Good suggestion, thanks, fixed!

janvorli commented on src/coreclr/build-test.cmd:
```diff
@@ -663,51 +663,59 @@ exit /b 0
+    set __CrossgenCmd=!__CrossgenExe! /Platform_Assemblies_Paths "!CORE_ROOT!" /in !AssemblyPath! /out !__CrossgenOutputFile! >nul 2>nul
 ) else (
-    "!CORE_ROOT_STAGE!\crossgen2\crossgen2 -r:"!CORE_ROOT!\*.dll" -O --inputbubble -out:"__CrossgenOutputFile" "%1" >nul 2>nul
+    set __CrossgenCmd=!__CrossgenExe! -r:"!CORE_ROOT!\*.dll" -O --inputbubble --out:!__CrossgenOutputFile! !AssemblyPath! >nul 2>nul
```
I suppose the `>nul 2>nul` was supposed to be part of the `__CrossgenCmd`, but it is applied to the `set` command instead. It should be placed on the `!__CrossgenCmd!` line if we want to suppress the crossgen output.

trylek replied:
Deleted per our offline chat, there's no sense in throwing away diagnostic information (and there's not a ton of it).

trylek committed:
Fix framework build with legacy Crossgen due to a typo in my change

trylek committed:
Addressing PR feedback

1) The ordering of steps in build-test was incorrect - we need to
patch CORE_ROOT with live-live libraries build before Crossgenning
the framework therein, otherwise weird things happen (and the result
is incorrect in any case).

2) Removed misplaced >nul 2>nul - we agreed with JanV that it's
better to not throw diagnostic information away.

3) Applied JanV's suggestion for improvement - using a tighter filter
for reference assemblies - Microsoft.*.dll / System.*.dll /
mscorlib.dll instead of just *.dll that ended up picking tons of
garbage like clrjit.dll and the various Win32 API contracts, both
in the build-test script and in CLRTest.Crossgen.targets.

Thanks

Tomas

janvorli reviewed (approved):
LGTM, thank you!
